server/util.go

```diff
@@ -157,3 +163,102 @@ func checkBootstrapRequest(clusterID uint64, req *pdpb.BootstrapRequest) error {
+
+// Serviceregistry used to install the registered services, including gRPC and HTTP API.
+type Serviceregistry interface {
+	InstallAllGRPCServices(srv *Server, g *grpc.Server)
+	InstallAllRESTHandler(srv *Server, userDefineHandler map[string]http.Handler)
+}
+
+// NewServiceregistry is a hook for msc code which implements the micro service.
+var NewServiceregistry = func() Serviceregistry {
+	return dummyServiceregistry{}
+}
+
+type dummyServiceregistry struct{}
+
+func (d dummyServiceregistry) InstallAllGRPCServices(srv *Server, g *grpc.Server) {
+}
+
+func (d dummyServiceregistry) InstallAllRESTHandler(srv *Server, userDefineHandler map[string]http.Handler) {
+}
+
+// APIServiceGroup used to register the HTTP REST API.
+type APIServiceGroup struct {
+	Name       string
+	Version    string
+	IsCore     bool
+	PathPrefix string
+}
+
+// Path returns the path of the service.
+func (sg *APIServiceGroup) Path() string {
+	if len(sg.PathPrefix) > 0 {
+		return sg.PathPrefix
+	}
+	if sg.IsCore {
+		return CorePath
+	}
+	if len(sg.Name) > 0 && len(sg.Version) > 0 {
+		return path.Join(ExtensionsPath, sg.Name, sg.Version)
+	}
+	return ""
+}
+
+// RegisterUserDefinedHandlers register the user defined handlers.
+func RegisterUserDefinedHandlers(registerMap map[string]http.Handler, group *APIServiceGroup, handler http.Handler) error {
+	pathPrefix := group.Path()
+	if _, ok := registerMap[pathPrefix]; ok {
+		return errs.ErrServiceRegistered.FastGenByArgs(pathPrefix)
+	}
+	if len(pathPrefix) == 0 {
+		return errs.ErrAPIInformationInvalid.FastGenByArgs(group.Name, group.Version)
+	}
+	registerMap[pathPrefix] = handler
+	log.Info("register REST path", zap.String("path", pathPrefix))
+	return nil
+}
+
+func combineBuilderServerHTTPService(ctx context.Context, svr *Server, serviceBuilders ...HandlerBuilder) (map[string]http.Handler, error) {
+	userHandlers := make(map[string]http.Handler)
+	registerMap := make(map[string]http.Handler)
+
+	apiService := negroni.New()
+	recovery := negroni.NewRecovery()
+	apiService.Use(recovery)
+	router := mux.NewRouter()
+
+	for _, build := range serviceBuilders {
+		handler, info, err := build(ctx, svr)
+		if err != nil {
+			return nil, err
+		}
+		if !info.IsCore && len(info.PathPrefix) == 0 && (len(info.Name) == 0 || len(info.Version) == 0) {
+			return nil, errs.ErrAPIInformationInvalid.FastGenByArgs(info.Name, info.Version)
+		}
+
+		if err := RegisterUserDefinedHandlers(registerMap, &info, handler); err != nil {
+			return nil, err
+		}
+	}
+
+	// Combine the pd service to the router. the extension service will be added to the userHandlers.
+	for pathPrefix, handler := range registerMap {
+		if strings.Contains(pathPrefix, CorePath) || strings.Contains(pathPrefix, ExtensionsPath) {
+			router.PathPrefix(pathPrefix).Handler(handler)
+			if pathPrefix == CorePath {
+				// Deprecated
+				router.Path("/pd/health").Handler(handler)
+				// Deprecated
+				router.Path("/pd/ping").Handler(handler)
```

Comment:
Do we still need to keep these?

Reply:
I'm not sure if there is still a dependency. we do not have relative test, so I keep it in here.